er description (#27)
SW705E19/report: sections/design/erd-diagram.tex

\section{Entity-relationship diagram}
An entity-relationship diagram (ER diagram) can express the structure of a database graphically\cite{DBConcepts}.

> This is a way to gain a better understanding of the requirements of the database and to ensure its design is clear.

Reviewer: ```suggestion
It serves as a way to gain a better understanding of the requirements of the database and to ensure its design is clear.
```

An ER diagram shows entity sets of the database and their relationships.

> An entity is a thing or an object in the real world that is distinguishable from other objects, and are described through a set of attributes.

Reviewer: ```suggestion
An entity is usually an object in the real world that is distinguishable from other objects, and is described through a set of attributes.
```

An entity set is a set of these entities.
Relationships are associations between entities.

> Cardinalities are important for these relationships - they express the number of entities to which another entity can be associated\cite{DBConcepts}, either one or many.

Reviewer: ```suggestion
Cardinalities are important for these relationships, as they express the number of entities to which another entity can be associated\cite{DBConcepts}, either one or many.
```

When an attribute is unique, it can be used to identify a given instance of the entity. 
However, because of the one-to-many relationship between service and material, material also contains what is known as a foreign key in the attribute \textit{ServiceID}.
This foreign key references the primary key of the entity set that participates as the one side. 
This ensures that for any entity in set \textit{material} you can use the foreign key to get the corresponding service.
\\\\
The relationship between \textit{service} and \textit{category} is a many-to-many relationship.
To model this, a separate entity set is created.
This entity set is named after the two participating sets, and has both primary keys as its attributes.
This ensures that for a particular entity in the set \textit{service}, it can be related to multiple entities of the set \textit{category}.
In the database, this would be achieved through multiple entities of the \textit{service\_category} being present with the same \textit{ServiceID} attributes, but each instance is linked to a different \textit{CategoryID} attribute.
Another example of a many-to-many relation is users and messages.

> A message is associated with a sender, and at least one receiver.

Reviewer: ```suggestion
Another example of a many-to-many relation is users and messages, where a message is associated with a sender, and at least one receiver.
```

\\\\
A particularly interesting entity set is the \textit{user} set. 
As defined in the \autoref{fig:class-diagram}, the admin and tutor properties are viewed as specializations through roles.
This is achieved in the ER diagram by modelling them as subtypes, defining an \textit{ISA} relationship.

> This is what the arrow from both \textit{tutor} and \textit{admin} represent.

Reviewer: ```suggestion
This is achieved in the ER diagram by modeling them as subtypes, defining an \textit{ISA} relationship, which is represented by the arrow from both \textit{tutor} and \textit{admin}.
```
